apps/web/src/core/components/hooks/useWindowAI.ts

```diff
@@ -18,9 +16,11 @@ export const initialMessages: ChatMessage[] = [
 
 export function useWindowAI(
   defaultMessages = initialMessages,
-  { cacheSize = 10, stream = false } = {}
+  { cacheSize = 10, stream = false, keep = 0, maxTokens = 2048 } = {}
 ) {
-  const [messages, setMessages] = useState<ChatMessage[]>(defaultMessages)
+  const messagesRef = useRef<ChatMessage[]>(defaultMessages)
```

Comment:
I'm surprised that this was necessary, given the messages state below. was the line that says `messagesRef.current.pop()` the only reason? everything else seems like a simple state change

Reply:
Hahah - react's render state mutation does not propagate down to the context of a function callback that was initialized at the first render. This means that if I want to use the hook's sendMessage inside a long-running event listener with a static (no rerendering) callback, such as xterm's onData, I will need it to persist its state without relying on a rerender cycle via a ref! - i.e, the ref store the underlying state (and a way for memoized function to access said state), whereas  the rendering state is used to render into react runtime - for chat UI/agent debugger etc...

Reply:
ok, I think i get it... wow